src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -3600,34 +3634,200 @@ int RGWRados::rewrite_obj(RGWBucketInfo& dest_bucket_info, const rgw_obj& obj, c
+  if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": get_obj_head_ioctx for \"" << obj << "\" returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
+  }
 
-  int ret = update_idx.prepare(dpp, RGWModifyOp::CLS_RGW_OP_ADD, no_write_tag, y);
+  const uint64_t epoch = head_obj_ctx.get_last_version();
+  const int64_t poolid = head_obj_ctx.get_id();
+
+  const bool is_versioned = bucket_info.versioned();
+  const bool has_instance = ! obj.key.instance.empty();
+
+  ldpp_dout(dpp, 20) << "INFO: " << __func__ << ": reindexing " <<
+    obj.key << " in bucket " << bucket_info.bucket << dendl;
+
+  // complete stat op
+  ret = stat_op.wait(dpp);
+  if (ret == -ENOENT && is_versioned && has_instance) {
+    // head object does not exist if it's a delete marker; handle here
+    // and return
+    ldpp_dout(dpp, 20) << "INFO: " << __func__ << ": indexing " <<
+      obj.key << " as delete marker" << dendl;
+    ret = set_olh(dpp,
+		  obj_ctx,
+		  bucket_info,
+		  obj,
+		  true, // is delete marker
+		  nullptr, // &metadata
+		  0, // uint64_t olh_epoch,
+		  ceph::real_clock::zero(), // unmod_since,
+		  true, // high_precision_time
+		  null_yield,
+		  nullptr, // zones_trace,
+		  false,  // log_data_change
+		  true); // skip the olh object updaste
+    if (ret < 0) {
+      ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+	": unable to set OLH to delete marker for \"" << obj << "\"; returned: " <<
+	cpp_strerror(-ret) << dendl;
+    }
+
+    return ret;
+  } else if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ <<
+      ": unable to complete async stat of \"" << obj << "\"; returned: " <<
+      cpp_strerror(-ret) << dendl;
+    return ret;
+  }
+
+  ceph::real_clock::time_point mtime = ceph::real_clock::from_timespec(stat_op.result.mtime);
+  bufferlist acl_bl;
+  std::string etag;
+  std::string content_type;
+  std::string storage_class;
+  bool found_compressed;
+  bufferlist compressed_bl;
+  bool found_manifest = false;
+  bufferlist manifest_bl;
+  bool appendable = false;
+  bufferlist part_num_bl;
+
+  {
+    auto& attrs = stat_op.result.attrs;
+    read_attr(attrs, RGW_ATTR_ACL, std::move(acl_bl));
+    read_attr(attrs, RGW_ATTR_ETAG, etag);
+    read_attr(attrs, RGW_ATTR_CONTENT_TYPE, content_type);
+    read_attr(attrs, RGW_ATTR_STORAGE_CLASS, storage_class);
+    read_attr(attrs, RGW_ATTR_COMPRESSION, std::move(compressed_bl), &found_compressed);
+    read_attr(attrs, RGW_ATTR_MANIFEST, std::move(manifest_bl), &found_manifest);
+    read_attr(attrs, RGW_ATTR_APPEND_PART_NUM, std::move(part_num_bl), &appendable);
+  }
+
+  uint64_t stored_size = stat_op.result.size;    // size consumed in cluster
+  uint64_t accounted_size = stat_op.result.size; // size if/when decompressed
+
+  if (!found_manifest) {
+    // this appears to be a pure OLH (i.e., no data); we won't index
+    // as we index actual objects with data and set the OLH then
```

Comment:
presence of `RGW_ATTR_OLH_VER` is probably a better way to detect this case. very old objects may not have a manifest

Reply:
Thanks!